### PR TITLE
Fix slither for main contracts

### PR DESCRIPTION
**Pull request by kronosapiens**
Partially solves #319

Updates our `slither` invocation for the main contracts, to be included in Cerulean LWSS. Does not include the tooling changes (adding to Circle) and updates to extensions, which can be found in #929 

As a nice perk, we can ditch `slither.db` since slither now allows for inline exemptions.

### CODE REVIEW

**Comment by area** [comment]
Confidence is being dulled here by `ERROR:ContractSolcParsing:Missing function 'expression'`, and I can't figure out what's causing it...

**Comment by area** [comment]
Okay, well, ignoring that mysterious error, as far as I can see the issues it's raising fall in to a few categories:

* False-raising of `unused-return` in `try-catch` blocks.
* A few legitimate raisings of `unused-return`. We probably do want to check these are behaving in ways we are happy with in situations where `false` is returned, rather than the calls just reverting, in the same way we accommodated that behaviour in `TokenLocking`. The one in ColonyNetworkMining is okay, I think, because it's targeting CLNY, but the one in CoinMachine does probably need to be in a `require`.
* 'Dangerous' strict inequalities where, as far as I can tell, they are appropriate.
* Test helper contracts, which can be ignored.
* `push`ing on to arrays it doesn't like, for some reason. It's claiming that the user is able to set the length, which I don't think is the case (beyond adding elements one-by-one).
* `selfdestruct` having the gall to exit in the mining contracts.

**Comment by kronosapiens** [comment]
I've addresses all the warnings so slither's output is now clean. To Alex's point, we might want to think more about how to handle the "unused return" values for token transfers -- I've simply thrown `require`s around all of them for now, which may not be what we want.

Still not sure what the `Missing function: 'expression'` warning means...

**Comment by kronosapiens** [comment]
Ok, this is in reasonable shape for a review. I did some (I believe) no-op code reorganization to squash some reentrancy warnings, but please tell me if it looks like I've done something terrible. I also took some of the strict equality warnings and made them inequalities even though I'm pretty sure there's logically no way for the inequality to ever occur. If you feel like that's a misleading or confusing change I'd be happy to put back the equalities and exclude the warning instead.

**Review by area — COMMENTED** [review]
How much stuff pops up if we allow low severity warnings? Suuuuuuper glad to see that we can use comments to ignore stuff now though, that's really good to see.

File: contracts/colonyNetwork/ColonyNetwork.sol [hunk 1]
```diff
@@ -244,6 +244,7 @@ contract ColonyNetwork is ColonyNetworkStorage, MultiChain {
         // When we are at an integer power of two steps away from the newly added skill (leaf) node,
         // add the current parent skill to the new skill's parents array
         if (treeWalkingCounter == powerOfTwo) {
+          // slither-disable-next-line controlled-array-length
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply contracts/colonyNetwork/ColonyNetwork.sol:247]: For some reason `push` seems to trigger the warning around setting array length, even though it seems pretty clearly different.

File: contracts/colonyNetwork/ColonyNetworkMining.sol [hunk 2]
```diff
@@ -249,9 +252,9 @@ contract ColonyNetworkMining is ColonyNetworkStorage, MultiChain {
...
     ITokenLocking(tokenLocking).withdraw(clnyToken, _amount, true);
     if (isXdai()){
-      // On Xdai, I'm burning bridged tokens is certainly not what we want. So let's
-      // send them to the metacolony for now.
-      ERC20Extended(clnyToken).transfer(metaColony, _amount);
+      // On Xdai, I'm burning bridged tokens is certainly not what we want.
+      //   So let's send them to the metacolony for now.
+      require(ERC20Extended(clnyToken).transfer(metaColony, _amount), "colony-network-transfer-failed");
     } else {
       ERC20Extended(clnyToken).burn(_amount);
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply contracts/colonyNetwork/ColonyNetworkMining.sol:259]: In the ERC20 spec, `burn` doesn't have a return value!

**Comment by kronosapiens** [comment]
@area I didn't look at the low-severity warnings, I'll do that today.

**Comment by kronosapiens** [comment]
@area I took a look at the low-severity warnings. There are quite a few of them but they are, unexpectedly, not severe. I added some new events and renamed some variables to clear some of the warnings but on balance I think we're better off just excluding all low-severity warnings rather than addressing them or excluding them on a per-detector basis.

**Comment by area** [comment]
Yeah, that's fair enough. We clearly thought the same in the past as well, but worth checking again!

File: contracts/colonyNetwork/ColonyNetworkMining.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (area) [reply contracts/colonyNetwork/ColonyNetworkMining.sol:259]: Ah, of course! It's smarter than I give it credit for :smile:

**Comment by area** [comment]
Why the new events?

**Comment by kronosapiens** [comment]
@area it was one of slither's warnings -- it didn't like storage variables being assigned without some sort of fanfare.

https://github.com/crytic/slither/wiki/Detector-Documentation#missing-events-access-control

https://github.com/crytic/slither/wiki/Detector-Documentation#missing-events-arithmetic

**Comment by area** [comment]
Let's start running slither as part of our builds with this PR, just excluding the extension directory.

**Comment by kronosapiens** [comment]
@area fair, I've pulled that commit in and excluded the `extensions/` dir.

**Review by area — APPROVED** [review]

**Comment by area** [comment]
Merged with superpowers because... GitHub was being a doofus?

**Comment by kronosapiens** [comment]
Yeah I was wondering what that was...

**Pull request merged**